packages/wrangler/src/pages/publish.tsx

```diff
@@ -327,47 +338,66 @@ export const Handler = async ({
 	}
 
 	if (builtFunctions) {
+		// with Pages Functions
+		// https://developers.cloudflare.com/pages/platform/functions/
 		formData.append("_worker.js", new File([builtFunctions], "_worker.js"));
 		logger.log(`✨ Uploading Functions`);
-		try {
-			_routes = readFileSync(routesOutputPath, "utf-8");
-			if (_routes) {
+
+		if (_routes) {
+			// with custom _routes.json file
+			try {
+				_routes = validateAndOptimizeRoutesFile(
+					_routes,
+					join(directory, "_routes.json")
+				);
+
 				formData.append("_routes.json", new File([_routes], "_routes.json"));
+				logger.log(`✨ Uploading _routes.json`);
+				logger.warn(
+					`_routes.json is an experimental feature and is subject to change. Please use with care.`
+				);
+			} catch (err) {
+				// forward the potential FatalError from an invalid spec
+				if (err instanceof FatalError) {
+					throw err;
+				}
 			}
-		} catch {}
+		} else {
+			// with generated _routes.json file
+			try {
+				_routes = readFileSync(
+					routesOutputPath as PathOrFileDescriptor,
```

Comment:
Are we casting this because `routesOutputPath` might be undefined? If so, I think it might be cleaner to do `else if(routesOutputPath !== undefined){}` on line 365 rather than casting